nova/src/folding/hypernova/cyclefold/nimfs/mod.rs

```diff
@@ -0,0 +1,330 @@
+            (RelaxedR1CSInstance<G2, C2>, RelaxedR1CSWitness<G2>),
+        ),
+        Error,
+    > {
+        let mut random_oracle = RO::new(config);
+
+        random_oracle.absorb(&vk);
+        random_oracle.absorb(&U);
+        random_oracle.absorb(&u);
+        random_oracle.absorb_non_native(&U_secondary);
+
+        let rho: G1::BaseField =
+            random_oracle.squeeze_field_elements_with_sizes(&[SQUEEZE_ELEMENTS_BIT_SIZE])[0];
+        let rho_scalar: G1::ScalarField =
+            unsafe { cast_field_element::<G1::BaseField, G1::ScalarField>(&rho) };
+
+        let (hypernova_proof, (folded_U, folded_W)) =
+            HNProof::prove_as_subprotocol(&mut random_oracle, shape, (U, W), (u, w), &rho_scalar)?;
+
+        // The PolyCommitment trait only guarantees the commitment can be represented by a vector of field elements. However,
+        // it makes sense to implement HyperNova Cyclefold assuming the commitment is just a single field element, because we
+        // will be using such a scheme for now (Zeromorph) and it leads to a minimally-sized secondary circuit. We expect this
+        // conversion will panic as unimplemented!() for any incompatible such commitment scheme.
+
+        let s_U_commitment_W = U.commitment_W.clone().into_single();
+        let s_u_commitment_W = u.commitment_W.clone().into_single();
+
+        let g_out = s_U_commitment_W + s_u_commitment_W * rho_scalar;
+        let W_comm_trace = secondary::synthesize::<G1, G2, C2>(
+            secondary::Circuit {
+                g1: s_U_commitment_W,
+                g2: s_u_commitment_W,
+                g_out,
+                r: rho,
+            },
+            pp_secondary,
+        )?;
+        debug_assert!(shape_secondary
+            .is_satisfied(&W_comm_trace.0, &W_comm_trace.1, pp_secondary)
+            .is_ok());
+
+        let (T, commitment_T) = r1cs::commit_T(
+            shape_secondary,
+            pp_secondary,
+            U_secondary,
+            W_secondary,
+            &W_comm_trace.0,
+            &W_comm_trace.1,
+        )?;
+        random_oracle.absorb_non_native(&W_comm_trace.0);
+        random_oracle.absorb(&commitment_T.into_affine());
+        random_oracle.absorb(&cast_field_element_unique::<G1::BaseField, G1::ScalarField>(&rho));
+
+        let rho_p: G1::BaseField =
+            random_oracle.squeeze_field_elements_with_sizes(&[SQUEEZE_ELEMENTS_BIT_SIZE])[0];
+
+        let U_secondary = U_secondary.fold(&W_comm_trace.0, &commitment_T, &rho_p)?;
+        let W_secondary = W_secondary.fold(&W_comm_trace.1, &T, &rho_p)?;
+
+        let commitment_W_proof = secondary::Proof { commitment_T, U: W_comm_trace.0 };
+
+        let proof = Self {
+            commitment_T,
+            commitment_W_proof,
+            hypernova_proof,
+            _poly_commitment: PhantomData,
+        };
+
+        Ok((proof, (folded_U, folded_W), (U_secondary, W_secondary)))
+    }
+
+    #[cfg(any(test, feature = "spartan"))]
```

Comment:
This is probably a copy-paste.

I suggest we vote whether we want to feature-gate anything. `spartan` is defaulted everywhere in the workspace, and disabling it will break hypernova because of traits and zeromorph dependencies. I proposed adding `feature = "spartan"` as I thought people might want to opt out from it (microsoft Nova e.g. doesn't care).
So either we handle it properly or drop it for the time being.

@sjudson @dorebell wdyt?

(this PR can do whatever compiles **and** correct w.r.t. hypernova paper, so we don't block it from merging with this discussion)

Reply:
Yeah it's a copy-paste. I don't have a strong opinion about whether to keep it gated: the two of you have a much better sense of the intended evolution of the repo than I do at present.